allow creating a prod op with a qfunc

timmysilv opened the pull request:
**Context:**
Quantum functions (aka qfuncs) are helpful ways of encapsulating a collection of operators. So is the new Prod operator!

**Description of the Change:**
Users can pass a qfunc to `qml.prod` in order to effectively convert from a qfunc to a Prod

**Benefits:**
qfuncs are loosely defined, since they are vanilla python callables and PennyLane can't do much with them. Prods are PennyLane operators and generally much more powerful objects. This will hopefully make stuff easier for users as well.

**Possible Drawbacks:**
Errors/behaviour can get uglier when mis-used, but I don't think it's worth the extra type-checking (lmk if you want it though). Here are some examples to open discussion:
1. The wire order might not be what the user expects because we reverse the order of operators to create the Prod:
```python
def qfunc():
    qml.PauliX(0)
    qml.PauliY(1)

op = qml.prod(qfunc)()
print(op, op.wires)
# (PauliY(wires=[1]) @ PauliX(wires=[0]), <Wires = [1, 0]>)
```
2. This raises `AttributeError: 'function' object has no attribute 'wires'` because we can't combine function arguments with operators:
```python
def qfunc():
    qml.PauliX(0)
    qml.PauliY(1)

qml.prod(qfunc, qml.PauliZ(2))
```
3. `ParametrizedEvolution` is an Operator but it's also a callable, so `prod` would confuse it as a qfunc if passed on its own (which used to fail saying you need multiple operands), and now it just returns another wrapper whose behaviour honestly confuses me:
```python
f1 = lambda p, t: np.sin(p * t)
H = f1 * qml.PauliY(0)
ev = qml.evolve(H)
qml.prod(ev)([[1.1, 2.2], 2], t=2.)
# <function pennylane.pulse.parametrized_evolution.prod.<locals>.wrapper(params, t, return_intermediate=None, complementary=None, **odeint_kwargs)>
```
4. The behaviour isn't _quite_ like `qml.adjoint` because it can accept an operator or a function. `qml.adjoint(qfunc())` also fails like `prod`, but `qml.adjoint(some_op())` doesn't fail. There's no analog for this with `prod`.
```pycon
>>> print(qml.adjoint(qml.RX(1.1, wires=0)))
Adjoint(RX(1.1, wires=[0]))
>>> print(qml.adjoint(qml.RX)(1.1, wires=0))  # should this syntax.. not be used? is this an unintended side-effect?
Adjoint(RX(1.1, wires=[0]))
>>> print(qml.prod(qfunc)())
PauliY(wires=[1]) @ PauliX(wires=[0])
>>> print(qml.prod(qfunc()))
# ValueError: Require at least two operators to combine; got 1
```

[here](https://xanaduai-pennylane--4011.com.readthedocs.build/en/4011/code/api/pennylane.prod.html)'s the link to page in docs if you wanna see the example rendered

timmysilv commented:
[sc-37068]

albi3ro reviewed (commented):
Bringing in @trbromley for this discussion. What do we think about the syntax:
```
qml.prod(qfunc)(*args, **kwargs)
```
Yes in this case we will return a function instead of a `Prod` operator, but we do support this behaviour in other places already, like `adjoint`.

If we don't want to have the same function with different arguments and outputs, maybe we could just create a new function `qml.qfunc_to_opeator`, or something else better named?

trbromley commented:
> Bringing in @trbromley for this discussion. What do we think about the syntax:
> 
> ```
> qml.prod(qfunc)(*args, **kwargs)
> ```
> 
> Yes in this case we will return a function instead of a `Prod` operator, but we do support this behaviour in other places already, like `adjoint`.
> 
> If we don't want to have the same function with different arguments and outputs, maybe we could just create a new function `qml.qfunc_to_opeator`, or something else better named?

+1 to having `qml.prod()` accept a `qfunc` and return another callable that returns a `Prod` when called.

@DSGuala @Jaybsoni will this help solve the issue with using `qfunc`s in `qml.QSVT`?

timmysilv commented:
sounds good. I've already updated the PR to match that, but I also updated the 4 cases of concern I thought of. One highlights why this isn't the same as `adjoint`. In short, `prod(qfunc)()` works, but `prod(qfunc())` does not. `adjoint` handles those differences nicely.

albi3ro commented on pennylane/ops/op_math/prod.py:
```diff
@@ -84,7 +85,26 @@ def prod(*ops, do_queue=True, id=None, lazy=True):
+
+    You can also create a prod operator by passing a qfunc to prod, like the following:
+
+    >>> def qfunc():
+    ...     qml.Hadamard(0)
+    ...     qml.CNOT([0, 1])
+    >>> prod_op = prod(qfunc)
+    >>> prod_op
+    CNOT(wires=[0, 1]) @ Hadamard(wires=[0])
```
Documentation example now out of date.

albi3ro commented on pennylane/ops/op_math/prod.py:
```diff
@@ -84,7 +85,26 @@ def prod(*ops, do_queue=True, id=None, lazy=True):
+
+        @wraps(fn)
+        def wrapper(*args, **kwargs):
+            qs = qml.tape.make_qscript(fn)(*args, **kwargs)
```
What behavior do we want if the qfunc only contains one object?

Since this will essentially be a general purpose qfunc to operator converter, I think its a valid situation that will occur. I'd be happy with just returning that individual op.

Though this does make me wonder if we should have some standalone function other than `qml.prod` as an "qfunc to operator" converter, since `qml.prod` means "product of multiple operators".

How do we *define* the error from a single operator qfunc out of existence?

timmysilv replied:
imo it currently fails as expected (saying that a prod must have more than one arg). I agree that if we want to support it, that probably belongs in a standalone function

DSGuala replied:
> > Bringing in @trbromley for this discussion. What do we think about the syntax:
> > ```
> > qml.prod(qfunc)(*args, **kwargs)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Yes in this case we will return a function instead of a `Prod` operator, but we do support this behaviour in other places already, like `adjoint`.
> > If we don't want to have the same function with different arguments and outputs, maybe we could just create a new function `qml.qfunc_to_opeator`, or something else better named?
> 
> +1 to having `qml.prod()` accept a `qfunc` and return another callable that returns a `Prod` when called.
> 
> @DSGuala @Jaybsoni will this help solve the issue with using `qfunc`s in `qml.QSVT`?

This will definitely make it easier to define custom `BlockEncoding` and `PCPhase` operations in `qml.QSVT`. I gave it a quick try and found it helpful :+1: 

Using `qml.prod` is still an extra step over submitting qfuncs directly to `qml.QSVT`. But it's much better than having to create a loop using `qml.prod` to get all the operations into one `qml.prod`.

We could potentially have `qml.QSVT` check the type the operations submitted to it and if they are qfuncs, automatically use `qml.prod` under the hood.

timmysilv commented:
the behaviour surrounding single-op qfuncs (and single-operand Prods in general) is too varied, and this concept (converting from some input type to a single operator) needs more consideration. We will pick this up in a new manner for the next release

albi3ro reviewed (approved):
I like this solution and think it will end up being very useful.

mudit2812 reviewed (approved):
Looks good to me
